### PR TITLE
Kinetica exporter/pr 2 metrics

### PR DESCRIPTION
**Pull request by am-kinetica**
**Description:** Added metrics handling

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24294

**Testing:** No new tests

**Documentation:** No new documentation

### CODE REVIEW

**Comment by am-kinetica** [comment]
Merged with changes from main

File: exporter/kineticaexporter/config.go [hunk 1]
```diff
@@ -34,7 +53,196 @@ func (cfg *Config) Validate() error {
...
+	return fallback
+}
+
+// createLogger - creates a Logger using user defined configuration from a file.
+// If the file name is not found or there is some error it will create a default
+// logger using the default logger config file named 'config_log_zap.yaml'.
+//
+//	@receiver cfg
+//	@return *zap.Logger
+func (cfg *Config) createLogger() *zap.Logger {
```
Comment (TylerHelmuth) [inline exporter/kineticaexporter/config.go:77]: Unless it is required for exporting, we should not create a specific `zap.Logger`.  Instead, we should use the logger supplied in `set exporter.CreateSettings`.

Comment (am-kinetica) [reply exporter/kineticaexporter/config.go:77]: What if someone wants to customize file based logging with rollover, logging formats etc. ?

Comment (TylerHelmuth) [reply exporter/kineticaexporter/config.go:77]: At the moment the collector doesn't support that. It does expose these options to users for its logger: https://github.com/open-telemetry/opentelemetry-collector/blob/7c5ecef11dff4ce5501c9683b277a25a61ea0f1a/service/telemetry/config.go#L33

**Review by TylerHelmuth — CHANGES_REQUESTED** [review]

File: exporter/kineticaexporter/config.go (hunk 1 above)
Comment (am-kinetica) [reply exporter/kineticaexporter/config.go:77]: I am working on removing this.

**Review by TylerHelmuth — COMMENTED** [review]
@am-kinetica any unit tests you can add would be good.

File: exporter/kineticaexporter/metrics_exporter.go [hunk 2]
```diff
@@ -5,26 +5,242 @@ package kineticaexporter // import "github.com/open-telemetry/opentelemetry-coll
...
 type kineticaMetricsExporter struct {
 	logger *zap.Logger
+	writer *KiWriter
 }
 
-func newMetricsExporter(_ *zap.Logger, _ *Config) *kineticaMetricsExporter {
+func newMetricsExporter(logger *zap.Logger, cfg *Config) *kineticaMetricsExporter {
+	if err := cfg.Validate(); err != nil {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (am-kinetica) [reply exporter/kineticaexporter/metrics_exporter.go:138]: Removed superfluous call to validate

File: exporter/kineticaexporter/common.go [hunk 3]
```diff
@@ -0,0 +1,804 @@
...
+type AggregationTemporality int
+
+// const - Metrics
+//
+//	@param AggregationTemporalityUnspecified
+const (
+	AggregationTemporalityUnspecified AggregationTemporality = iota
+	AggregationTemporalityDelta
+	AggregationTemporalityCumulative
+)
```
[the remark this answers is not in the document]
Comment (am-kinetica) [reply exporter/kineticaexporter/common.go:596]: Used the built-in type as suggested

File: cmd/checkapi/allowlist.txt [hunk 4]
```diff
@@ -1,5 +1,7 @@
 connector/servicegraphconnector
 exporter/kafkaexporter
+exporter/kineticaexporter
+exporter/pulsarexporter
```
[the remark this answers is not in the document]
Comment (am-kinetica) [reply cmd/checkapi/allowlist.txt:4]: Removing this results in this error `2023/10/13 09:23:49 exporter/kineticaexporter has more than one function: "AttributeValueToKineticaFieldValue,ChunkBySize,GetGpuDbInst,NewAttributeValue,NewFactory,NewKiWriter,ValidateStruct"
`; can't figure out why

[the remark this answers is not in the document]
Comment (am-kinetica) [reply cmd/checkapi/allowlist.txt:4]: I don't think there are duplicates anywhere.

[the remark this answers is not in the document]
Comment (am-kinetica) [reply cmd/checkapi/allowlist.txt:4]: e.g., ChunkBySize occurs only once in `common.go`

[the remark this answers is not in the document]
Comment (am-kinetica) [reply cmd/checkapi/allowlist.txt]: Modified the code to expose only `NewFactory` and nothing else.

**Review by TylerHelmuth — COMMENTED** [review]
Looks like there is another conflict to handle then we'll be good to go.

**Review by atoulme — APPROVED** [review]

**Comment by atoulme** [comment]
I think this PR looks good now. @TylerHelmuth mind taking another look?

**Review by TylerHelmuth — CHANGES_REQUESTED** [review]
@atoulme we're waiting on some unit tests for `metrics_exporter.go`.

File: exporter/kineticaexporter/exporter_metric_test.go [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
...a\x16primary_key\x12shard_key\x08uuid\x00$metric_description\x06\x08data\x0echar256\x10nullable\x00\x16metric_name\x04\x08data\x0echar256\x00\x16metric_unit\x06\x08data\x0echar256\x10nullable\x00\x00\x00\x028\"attribute_indexes\x00 collection_names\x08otel$compressed_columns\x000datasource_subscriptions\x00\x18foreign_keys\x00\"foreign_shard_key\x00$global_access_mode\x14read_write,is_automatic_partition\nfalse\x10is_dirty\x00\"is_view_persisted\x00\"last_refresh_time\x00(owner_resource_group<kinetica_system_resource_group*partition_definitions\x004partition_definitions_json\x04{}\x1cpartition_keys\x00\x1cpartition_type\x08NONE\x18record_bytes\x06784\x1crefresh_method\x00&remaining_table_ttl\x04-1\"request_avro_json\xda\x02{\"table_name\":\"otel.metric_gauge\",\"type_id\":\"303100960528700881\",\"options\":{\"foreign_keys\":\"\",\"is_replicated\":\"false\",\"is_result_table\":\"false\",\"no_error_if_exists\":\"true\"}}\"request_avro_type\x10is_table\x16schema_name\x08otel&strategy_definitionR( ( VRAM 1, RAM 5, DISK0 5, PERSIST 5 ) )\x1atable_monitor\x04{}\x12table_ttl\x04-1\x16total_bytes\x020\x1euser_chunk_size\x0e8000000\x1eview_table_name\x00\x00\x00\x00\x00\x00\x01\x01\x00\x00"
```
Comment (TylerHelmuth) [inline exporter/kineticaexporter/exporter_metric_test.go:26]: Can these commented-out vars be removed?

**Review by TylerHelmuth — COMMENTED** [review]
@am-kinetica wanted to chime in an say I'm still keeping an eye on this PR, excited to see the progress made on the unit tests.

File: exporter/kineticaexporter/exporter_metric_test.go (hunk 5 above)
Comment (am-kinetica) [reply exporter/kineticaexporter/exporter_metric_test.go:26]: Yes, will remove them now that the tests have run

Comment (am-kinetica) [reply exporter/kineticaexporter/exporter_metric_test.go]: Removed the commented vars from the test.

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**